

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Cloud URL Examples &mdash; Bed Reader 1.0.6 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=e2a723ec"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="bed_reader Documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Bed Reader
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Cloud URL Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#http-section">Http</a></li>
<li class="toctree-l2"><a class="reference internal" href="#local-file-section">Local File</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#local-file-url">Local File URL</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#aws-section">AWS S3</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Bed Reader</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content style-external-links">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Cloud URL Examples</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/cloud_urls.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cloud-url-examples">
<h1>Cloud URL Examples<a class="headerlink" href="#cloud-url-examples" title="Link to this heading"></a></h1>
<p><em>Table of Contents</em>:</p>
<ul class="simple">
<li><p><a class="reference external" href="#http-section">Http</a></p></li>
<li><p><a class="reference external" href="#local-file-section">local file</a></p></li>
<li><p><a class="reference external" href="#aws-section">AWS S3</a></p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <cite>bed-reader</cite> package also supports Azure and GCP, but we don’t have examples.</p>
</div>
<p>To specify a file in the cloud, you must specify URL string plus optional cloud options.</p>
<p>The exact details depend on the cloud service. We’ll look at <cite>http</cite>, at <cite>local files</cite>, and at <cite>AWS S3</cite>.</p>
<section id="http-section">
<span id="id1"></span><h2>Http<a class="headerlink" href="#http-section" title="Link to this heading"></a></h2>
<p>You can read *.bed files from web sites directly. For small files, access will be fast. For medium-sized files,
you may need to extend the default <cite>timeout</cite>.</p>
<p>Reading from large files can also be practical and even fast under these conditions:</p>
<ul class="simple">
<li><p>You need only some of the information</p></li>
<li><p>(Optional, but helpful) You can provide some metadata about individuals (samples) and SNPs (variants) locally.</p></li>
</ul>
<p>Let’s first look at reading a small or medium-sized dataset.</p>
<p><em>Example:</em></p>
<p>Read an entire file and find the fraction of missing values.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bed_reader</span><span class="w"> </span><span class="kn">import</span> <span class="n">open_bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="s2">&quot;https://raw.githubusercontent.com/fastlmm/bed-sample-files/main/small.bed&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">missing_count</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">val</span><span class="p">)</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">missing_fraction</span> <span class="o">=</span> <span class="n">missing_count</span> <span class="o">/</span> <span class="n">val</span><span class="o">.</span><span class="n">size</span>
<span class="gp">... </span>    <span class="n">missing_fraction</span>
<span class="go">np.float64(0.16666666666666666)</span>
</pre></div>
</div>
<p>When reading a medium-sized file, you may need to set a <cite>timeout</cite> in your cloud options. With a <cite>timeout</cite>,
you can give your code more than the default 30 seconds to read metadata from the *.fam and *.bim files
(or genomic data from *.bed).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>See <a class="reference external" href="https://docs.rs/object_store/latest/object_store/enum.ClientConfigKey.html">ClientConfigKey</a>
for a list of cloud options, such as <cite>timeout</cite>, that you can always use.</p>
</div>
<p>You may also wish to use <cite>.skip_format_check=True</cite> to avoid a fast,
early check of the *.bed file’s header.</p>
<p>Here we print the first five iids (individual or sample ids) and first five sids (SNP or variant ids).
We then, print all unique chromosome values. Finally, we read all data from chromosome 5 and print its dimensions.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bed_reader</span><span class="w"> </span><span class="kn">import</span> <span class="n">open_bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span>
<span class="gp">... </span>    <span class="s2">&quot;https://raw.githubusercontent.com/fastlmm/bed-sample-files/main/toydata.5chrom.bed&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">cloud_options</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;timeout&quot;</span><span class="p">:</span> <span class="s2">&quot;100s&quot;</span><span class="p">},</span>
<span class="gp">... </span>    <span class="n">skip_format_check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">bed</span><span class="o">.</span><span class="n">iid</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
<span class="gp">... </span>    <span class="n">bed</span><span class="o">.</span><span class="n">sid</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span>
<span class="gp">... </span>    <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span> <span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span> <span class="o">==</span> <span class="s2">&quot;5&quot;</span><span class="p">])</span>
<span class="gp">... </span>    <span class="n">val</span><span class="o">.</span><span class="n">shape</span>
<span class="go">array([&#39;per0&#39;, &#39;per1&#39;, &#39;per2&#39;, &#39;per3&#39;, &#39;per4&#39;], dtype=&#39;&lt;U11&#39;)</span>
<span class="go">array([&#39;null_0&#39;, &#39;null_1&#39;, &#39;null_2&#39;, &#39;null_3&#39;, &#39;null_4&#39;], dtype=&#39;&lt;U9&#39;)</span>
<span class="go">array([&#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;], dtype=&#39;&lt;U9&#39;)</span>
<span class="go">(500, 440)</span>
</pre></div>
</div>
<p>Now, let’s read from a large file containing data from over 1 million individuals (samples) and over 300,000 SNPs (variants). The file size is 91 GB. In this example, we read data for just one SNP (variant). If we know the number of individuals (samples) and SNPs (variants) exactly, we can read this SNP quickly and with just one file access.</p>
<p>What is the mean value of the SNP (variant) at index position 100,000?</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bed_reader</span><span class="w"> </span><span class="kn">import</span> <span class="n">open_bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span>
<span class="gp">... </span>    <span class="s2">&quot;https://www.ebi.ac.uk/biostudies/files/S-BSST936/genotypes/synthetic_v1_chr-10.bed&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">cloud_options</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;timeout&quot;</span><span class="p">:</span> <span class="s2">&quot;100s&quot;</span><span class="p">},</span>
<span class="gp">... </span>    <span class="n">skip_format_check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">iid_count</span><span class="o">=</span><span class="mi">1_008_000</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">sid_count</span><span class="o">=</span><span class="mi">361_561</span><span class="p">,</span>
<span class="gp">... </span>    <span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span> <span class="mi">100_000</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">val</span><span class="p">)</span>
<span class="go">np.float32(0.03391369)</span>
</pre></div>
</div>
<p>You can also download the *.fam and *.bim metadata files and then read from them locally while continuing to read the *.bed file from the cloud.
This gives you almost instant access to the metadata and the *.bed file. Here is an example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bed_reader</span><span class="w"> </span><span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># For this example, assume &#39;synthetic_v1_chr-10.fam&#39; and &#39;synthetic_v1_chr-10.bim&#39; are already downloaded</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># and &#39;local_fam_file&#39; and &#39;local_bim_file&#39; variables are set to their local file paths.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">local_fam_file</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;synthetic_v1_chr-10.fam&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">local_bim_file</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;synthetic_v1_chr-10.bim&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span>
<span class="gp">... </span>    <span class="s2">&quot;https://www.ebi.ac.uk/biostudies/files/S-BSST936/genotypes/synthetic_v1_chr-10.bed&quot;</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">fam_filepath</span><span class="o">=</span><span class="n">local_fam_file</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">bim_filepath</span><span class="o">=</span><span class="n">local_bim_file</span><span class="p">,</span>
<span class="gp">... </span>    <span class="n">skip_format_check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;iid_count=</span><span class="si">{</span><span class="n">bed</span><span class="o">.</span><span class="n">iid_count</span><span class="si">:</span><span class="s2">_</span><span class="si">}</span><span class="s2">, sid_count=</span><span class="si">{</span><span class="n">bed</span><span class="o">.</span><span class="n">sid_count</span><span class="si">:</span><span class="s2">_</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;iid=</span><span class="si">{</span><span class="n">bed</span><span class="o">.</span><span class="n">iid</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;sid=</span><span class="si">{</span><span class="n">bed</span><span class="o">.</span><span class="n">sid</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span><span class="si">}</span><span class="s2">...&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;unique chromosomes = </span><span class="si">{</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:</span><span class="mi">10</span><span class="p">,</span> <span class="p">::</span> <span class="n">bed</span><span class="o">.</span><span class="n">sid_count</span> <span class="o">//</span> <span class="mi">10</span><span class="p">])</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;val=</span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">iid_count=1_008_000, sid_count=361_561</span>
<span class="go">iid=[&#39;syn1&#39; &#39;syn2&#39; &#39;syn3&#39; &#39;syn4&#39; &#39;syn5&#39;]...</span>
<span class="go">sid=[&#39;chr10:10430:C:A&#39; &#39;chr10:10483:A:C&#39; &#39;chr10:10501:G:T&#39; &#39;chr10:10553:C:A&#39;</span>
<span class="go"> &#39;chr10:10645:G:A&#39;]...</span>
<span class="go">unique chromosomes = [&#39;10&#39;]</span>
<span class="go">val=[[0. 1. 0. 2. 0. 1. 0. 0. 0. 0. 0.]</span>
<span class="go"> [0. 0. 0. 0. 0. 0. 1. 1. 0. 0. 0.]</span>
<span class="go"> [0. 0. 2. 2. 0. 1. 0. 2. 0. 0. 0.]</span>
<span class="go"> [0. 1. 1. 1. 0. 0. 0. 0. 0. 0. 0.]</span>
<span class="go"> [0. 0. 1. 2. 0. 1. 0. 1. 0. 0. 0.]</span>
<span class="go"> [0. 0. 0. 0. 0. 1. 0. 1. 0. 0. 0.]</span>
<span class="go"> [0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]</span>
<span class="go"> [0. 0. 0. 2. 0. 0. 0. 1. 0. 0. 0.]</span>
<span class="go"> [0. 0. 0. 1. 0. 1. 0. 0. 0. 0. 0.]</span>
<span class="go"> [0. 0. 0. 2. 0. 1. 0. 1. 0. 0. 0.]]</span>
</pre></div>
</div>
</section>
<section id="local-file-section">
<span id="id2"></span><h2>Local File<a class="headerlink" href="#local-file-section" title="Link to this heading"></a></h2>
<p>We can specify a local file as if it is in the cloud. This is a great way to test cloud functions. For real work and better efficiency, however,
use the file’s path rather than its URL.</p>
<section id="local-file-url">
<h3>Local File URL<a class="headerlink" href="#local-file-url" title="Link to this heading"></a></h3>
<p>The URL for a local file takes the form <cite>file:///{encoded_file_name}</cite>. No cloud options are needed.</p>
<p><em>Example:</em></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">bed_reader</span><span class="w"> </span><span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">urllib.parse</span><span class="w"> </span><span class="kn">import</span> <span class="n">urljoin</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span><span class="w"> </span><span class="nn">pathlib</span><span class="w"> </span><span class="kn">import</span> <span class="n">Path</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;file name: </span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">file name: ...small.bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">url</span> <span class="o">=</span> <span class="n">urljoin</span><span class="p">(</span><span class="s2">&quot;file:&quot;</span><span class="p">,</span> <span class="n">Path</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span><span class="o">.</span><span class="n">as_uri</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;url: </span><span class="si">{</span><span class="n">url</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="go">url: file:///.../small.bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">url</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span> <span class="mi">2</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">val</span><span class="p">)</span>
<span class="go">[[nan]</span>
<span class="go"> [nan]</span>
<span class="go"> [ 2.]]</span>
</pre></div>
</div>
</section>
</section>
<section id="aws-section">
<span id="id3"></span><h2>AWS S3<a class="headerlink" href="#aws-section" title="Link to this heading"></a></h2>
<p>Let’s look next at reading a file (or part of a file) from AWS S3.</p>
<p>The URL for an AWS S3 file takes the form <cite>s3://{bucket_name}/{s3_path}</cite>.</p>
<p>AWS forbids putting some needed information in the URL. Instead, that information must go into a string-to-string
dictionary of cloud options. Specifically, we’ll put <cite>“aws_region”</cite>, <cite>“aws_access_key_id”</cite>, and <cite>“aws_secret_access_key”</cite> in
the cloud options.
For security, we pull the last two option values from a file rather than hard-coding them into the program.</p>
<p>See <a class="reference external" href="https://docs.rs/object_store/latest/object_store/enum.ClientConfigKey.html">ClientConfigKey</a> for a list of cloud options, such as <code class="docutils literal notranslate"><span class="pre">timeout</span></code>, that you can always use.
See <a class="reference external" href="https://docs.rs/object_store/latest/object_store/aws/enum.AmazonS3ConfigKey.html">AmazonS3ConfigKey</a> for a list of AWS-specific options.
See <a class="reference external" href="https://docs.rs/object_store/latest/object_store/azure/enum.AzureConfigKey.html">AzureConfigKey</a> for a list of Azure-specific options.
See <a class="reference external" href="https://docs.rs/object_store/latest/object_store/gcp/enum.GoogleConfigKey.html">GoogleConfigKey</a> for a list of Google-specific options.</p>
<p><em>Example:</em></p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>I can run this, but others can’t because of the authentication checks.</p>
</div>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">configparser</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bed_reader</span><span class="w"> </span><span class="kn">import</span> <span class="n">open_bed</span>

<span class="n">config</span> <span class="o">=</span> <span class="n">configparser</span><span class="o">.</span><span class="n">ConfigParser</span><span class="p">()</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">config</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s2">&quot;~/.aws/credentials&quot;</span><span class="p">))</span>

<span class="n">cloud_options</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;aws_region&quot;</span><span class="p">:</span> <span class="s2">&quot;us-west-2&quot;</span><span class="p">,</span>
    <span class="s2">&quot;aws_access_key_id&quot;</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;default&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;aws_access_key_id&quot;</span><span class="p">),</span>
    <span class="s2">&quot;aws_secret_access_key&quot;</span><span class="p">:</span> <span class="n">config</span><span class="p">[</span><span class="s2">&quot;default&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;aws_secret_access_key&quot;</span><span class="p">),</span>
<span class="p">}</span>

<span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="s2">&quot;s3://bedreader/v1/toydata.5chrom.bed&quot;</span><span class="p">,</span> <span class="n">cloud_options</span><span class="o">=</span><span class="n">cloud_options</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
    <span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">val</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="c1"># Expected output: (500, 10000)</span>
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="bed_reader Documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Carl Kadie.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>