<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>bed_reader Documentation &mdash; Bed Reader 1.0.3 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/documentation_options.js?v=baaebd52"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Cloud URL Examples" href="cloud_urls.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            Bed Reader
          </a>
              <div class="version">
                1.0.3
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="cloud_urls.html">Cloud URL Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="cloud_urls.html#http-section">Http</a></li>
<li class="toctree-l2"><a class="reference internal" href="cloud_urls.html#local-file-section">Local File</a><ul>
<li class="toctree-l3"><a class="reference internal" href="cloud_urls.html#local-file-url">Local File URL</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="cloud_urls.html#aws-section">AWS S3</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Bed Reader</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content style-external-links">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active"><code class="xref py py-mod docutils literal notranslate"><span class="pre">bed_reader</span></code> Documentation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <a class="reference external image-reference" href="https://badge.fury.io/py/bed-reader"><img alt="https://badge.fury.io/py/bed-reader.svg" src="https://badge.fury.io/py/bed-reader.svg" /></a>
<a class="reference external image-reference" href="https://github.com/fastlmm/bed-reader/actions/workflows/ci.yml"><img alt="https://github.com/fastlmm/bed-reader/actions/workflows/ci.yml/badge.svg?branch=master" src="https://github.com/fastlmm/bed-reader/actions/workflows/ci.yml/badge.svg?branch=master" /></a>
<img alt="https://img.shields.io/pypi/pyversions/bed-reader" src="https://img.shields.io/pypi/pyversions/bed-reader" />
<section id="bed-reader-documentation">
<h1><code class="xref py py-mod docutils literal notranslate"><span class="pre">bed_reader</span></code> Documentation<a class="headerlink" href="#bed-reader-documentation" title="Link to this heading"></a></h1>
<p>Read and write the PLINK BED format, simply and efficiently.</p>
<p>Features:</p>
<ul class="simple">
<li><p>Fast multi-threaded Rust engine.</p></li>
<li><p>Supports all Python indexing methods. Slice data by individuals (samples) and/or SNPs (variants).</p></li>
<li><p>Used by <a class="reference external" href="https://github.com/fastlmm/PySnpTools">PySnpTools</a>, <a class="reference external" href="https://github.com/fastlmm/FaST-LMM">FaST-LMM</a>, and <a class="reference external" href="https://github.com/pystatgen">PyStatGen</a>.</p></li>
<li><p>Supports <a class="reference external" href="https://www.cog-genomics.org/plink2/formats">PLINK 1.9</a>.</p></li>
<li><p>Read data locally or from the cloud, efficiently and directly.</p></li>
</ul>
<section id="install">
<h2>Install<a class="headerlink" href="#install" title="Link to this heading"></a></h2>
<p><strong>Full version</strong>: With all optional dependencies:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>bed-reader<span class="o">[</span>samples,sparse<span class="o">]</span>
</pre></div>
</div>
<p><strong>Minimal version</strong>: Depends only on <cite>numpy</cite>:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>bed-reader
</pre></div>
</div>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h2>
<p>Read genotype data from a .bed file.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">val</span><span class="p">)</span>
<span class="go">[[ 1.  0. nan  0.]</span>
<span class="go"> [ 2.  0. nan  2.]</span>
<span class="go"> [ 0.  1.  2.  0.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">bed</span>
</pre></div>
</div>
<p>Read every second individual and SNPs (variants) from 20 to 30.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">file_name2</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;some_missing.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed2</span> <span class="o">=</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">val2</span> <span class="o">=</span> <span class="n">bed2</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[::</span><span class="mi">2</span><span class="p">,</span><span class="mi">20</span><span class="p">:</span><span class="mi">30</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">val2</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">(50, 10)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">bed2</span>
</pre></div>
</div>
<p>List the first 5 individual (sample) ids, the
first 5 SNP (variant) ids, and every unique
chromosome. Then, read every value in chromosome 5.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name2</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed3</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed3</span><span class="o">.</span><span class="n">iid</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed3</span><span class="o">.</span><span class="n">sid</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">bed3</span><span class="o">.</span><span class="n">chromosome</span><span class="p">))</span>
<span class="gp">... </span>    <span class="n">val3</span> <span class="o">=</span> <span class="n">bed3</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="n">bed3</span><span class="o">.</span><span class="n">chromosome</span><span class="o">==</span><span class="s1">&#39;5&#39;</span><span class="p">])</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">val3</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">[&#39;iid_0&#39; &#39;iid_1&#39; &#39;iid_2&#39; &#39;iid_3&#39; &#39;iid_4&#39;]</span>
<span class="go">[&#39;sid_0&#39; &#39;sid_1&#39; &#39;sid_2&#39; &#39;sid_3&#39; &#39;sid_4&#39;]</span>
<span class="go">[&#39;1&#39; &#39;10&#39; &#39;11&#39; &#39;12&#39; &#39;13&#39; &#39;14&#39; &#39;15&#39; &#39;16&#39; &#39;17&#39; &#39;18&#39; &#39;19&#39; &#39;2&#39; &#39;20&#39; &#39;21&#39; &#39;22&#39;</span>
<span class="go"> &#39;3&#39; &#39;4&#39; &#39;5&#39; &#39;6&#39; &#39;7&#39; &#39;8&#39; &#39;9&#39;]</span>
<span class="go">(100, 6)</span>
</pre></div>
</div>
<p>From the cloud: open a file and read data for one SNP (variant)
at index position 2. (See <a class="reference internal" href="cloud_urls.html"><span class="doc">Cloud URL Examples</span></a> for details on cloud URLs.)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="s2">&quot;https://raw.githubusercontent.com/fastlmm/bed-sample-files/main/small.bed&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">val</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="mi">2</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;float64&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">val</span><span class="p">)</span>
<span class="go">[[nan]</span>
<span class="go"> [nan]</span>
<span class="go"> [ 2.]]</span>
</pre></div>
</div>
</section>
<section id="project-links">
<h2>Project Links<a class="headerlink" href="#project-links" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://fastlmm.github.io/bed-reader">Documentation</a></p></li>
<li><p>Questions to <a class="reference external" href="mailto:fastlmm-dev&#37;&#52;&#48;python&#46;org">fastlmm-dev<span>&#64;</span>python<span>&#46;</span>org</a></p></li>
<li><p><a class="reference external" href="https://github.com/fastlmm/bed-reader">Source code</a></p></li>
<li><p><a class="reference external" href="https://pypi.org/project/bed-reader">PyPI</a></p></li>
<li><p><a class="reference external" href="https://github.com/fastlmm/bed-reader/issues">Bug reports</a></p></li>
<li><p><a class="reference external" href="https://mail.python.org/mailman3/lists/fastlmm-user.python.org">Mailing list</a></p></li>
<li><p><a class="reference external" href="https://fastlmm.github.io/">Project Website</a></p></li>
<li><p><a class="reference external" href="https://github.com/fastlmm/bed-reader/blob/master/CHANGELOG.md">Change Log</a></p></li>
</ul>
</section>
<section id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Link to this heading"></a></h2>
<section id="open-read-and-write">
<h3>Open, Read, and Write<a class="headerlink" href="#open-read-and-write" title="Link to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed" title="bed_reader.open_bed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed</span></code></a>(location[, iid_count, sid_count, ...])</p></td>
<td><p>Open a PLINK .bed file, local or cloud, for reading.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.read" title="bed_reader.open_bed.read"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.read</span></code></a>([index, dtype, order, ...])</p></td>
<td><p>Read genotype information.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.read_sparse" title="bed_reader.open_bed.read_sparse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.read_sparse</span></code></a>([index, dtype, ...])</p></td>
<td><p>Read genotype information into a <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/sparse.html#module-scipy.sparse" title="(in SciPy v1.13.0)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scipy.sparse</span></code></a> matrix.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.to_bed" title="bed_reader.to_bed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_bed</span></code></a>(filepath, val[, properties, ...])</p></td>
<td><p>Write values to a file in PLINK .bed format.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.create_bed" title="bed_reader.create_bed"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_bed</span></code></a>(location, iid_count, sid_count[, ...])</p></td>
<td><p>Open a file to write values into PLINK .bed format.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="properties-of-individuals-samples-and-snps-variants">
<h3>Properties of Individuals (samples) and SNPs (variants)<a class="headerlink" href="#properties-of-individuals-samples-and-snps-variants" title="Link to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.iid_count" title="bed_reader.open_bed.iid_count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.iid_count</span></code></a></p></td>
<td><p>Number of individuals (samples).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.sid_count" title="bed_reader.open_bed.sid_count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.sid_count</span></code></a></p></td>
<td><p>Number of SNPs (variants).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.shape" title="bed_reader.open_bed.shape"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.shape</span></code></a></p></td>
<td><p>Number of individuals (samples) and SNPs (variants).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.fid" title="bed_reader.open_bed.fid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.fid</span></code></a></p></td>
<td><p>Family id of each individual (sample).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.iid" title="bed_reader.open_bed.iid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.iid</span></code></a></p></td>
<td><p>Individual id of each individual (sample).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.father" title="bed_reader.open_bed.father"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.father</span></code></a></p></td>
<td><p>Father id of each individual (sample).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.mother" title="bed_reader.open_bed.mother"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.mother</span></code></a></p></td>
<td><p>Mother id of each individual (sample).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.sex" title="bed_reader.open_bed.sex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.sex</span></code></a></p></td>
<td><p>Sex of each individual (sample).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.pheno" title="bed_reader.open_bed.pheno"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.pheno</span></code></a></p></td>
<td><p>A phenotype for each individual (sample) (seldom used).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.sid" title="bed_reader.open_bed.sid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.sid</span></code></a></p></td>
<td><p>SNP id of each SNP (variant).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.chromosome" title="bed_reader.open_bed.chromosome"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.chromosome</span></code></a></p></td>
<td><p>Chromosome of each SNP (variant)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.cm_position" title="bed_reader.open_bed.cm_position"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.cm_position</span></code></a></p></td>
<td><p>Centimorgan position of each SNP (variant).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.bp_position" title="bed_reader.open_bed.bp_position"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.bp_position</span></code></a></p></td>
<td><p>Base-pair position of each SNP (variant).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.allele_1" title="bed_reader.open_bed.allele_1"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.allele_1</span></code></a></p></td>
<td><p>First allele of each SNP (variant).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.allele_2" title="bed_reader.open_bed.allele_2"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.allele_2</span></code></a></p></td>
<td><p>Second allele of each SNP (variant),</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.major" title="bed_reader.open_bed.major"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.major</span></code></a></p></td>
<td><p>Major mode of a local .bed file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.open_bed.properties" title="bed_reader.open_bed.properties"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.properties</span></code></a></p></td>
<td><p>All the properties returned as a dictionary.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.open_bed.property_item" title="bed_reader.open_bed.property_item"><code class="xref py py-obj docutils literal notranslate"><span class="pre">open_bed.property_item</span></code></a>(name)</p></td>
<td><p>Retrieve one property by name.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="utilities">
<h3>Utilities<a class="headerlink" href="#utilities" title="Link to this heading"></a></h3>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#bed_reader.sample_file" title="bed_reader.sample_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sample_file</span></code></a>(filepath)</p></td>
<td><p>Retrieve a sample .bed file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#bed_reader.tmp_path" title="bed_reader.tmp_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tmp_path</span></code></a>()</p></td>
<td><p>Return a <a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pathlib.Path</span></code></a> to a temporary directory.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="details">
<h2>Details<a class="headerlink" href="#details" title="Link to this heading"></a></h2>
<section id="open-bed">
<h3>open_bed<a class="headerlink" href="#open-bed" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="bed_reader.open_bed">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bed_reader.</span></span><span class="sig-name descname"><span class="pre">open_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/urllib.parse.html#urllib.parse.ParseResult" title="(in Python v3.12)"><span class="pre">ParseResult</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">iid_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sid_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">properties</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_A1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_format_check</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fam_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/urllib.parse.html#urllib.parse.ParseResult" title="(in Python v3.12)"><span class="pre">ParseResult</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bim_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/urllib.parse.html#urllib.parse.ParseResult" title="(in Python v3.12)"><span class="pre">ParseResult</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cloud_options</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_concurrent_requests</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_chunk_bytes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fam_filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bim_filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bed_reader/_open_bed.html#open_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.open_bed" title="Link to this definition"></a></dt>
<dd><p>Open a PLINK .bed file, local or cloud, for reading.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>location</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a>) – File path or URL to the .bed file. See <a class="reference internal" href="cloud_urls.html"><span class="doc">Cloud URL Examples</span></a> for details on cloud URLs.</p></li>
<li><p><strong>iid_count</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Number of individuals (samples) in the .bed file.
The default (<code class="docutils literal notranslate"><span class="pre">iid_count=None</span></code>) finds the number
automatically by quickly scanning the .fam file.</p></li>
<li><p><strong>sid_count</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Number of SNPs (variants) in the .bed file.
The default (<code class="docutils literal notranslate"><span class="pre">sid_count=None</span></code>) finds the number
automatically by quickly scanning the .bim file.</p></li>
<li><p><strong>properties</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a><em>, </em><em>optional</em>) – <p>A dictionary of any replacement properties. The default is an empty dictionary.
The keys of the dictionary are the names of the properties to replace.
The possible keys are:</p>
<blockquote>
<div><blockquote>
<div><p>”fid” (family id), “iid” (individual or sample id), “father” (father id),
“mother” (mother id), “sex”, “pheno” (phenotype), “chromosome”, “sid”
(SNP or variant id), “cm_position” (centimorgan position), “bp_position”
(base-pair position), “allele_1”, “allele_2”.</p>
</div></blockquote>
<p>The values are replacement lists or arrays. A value can also be <cite>None</cite>,
meaning do not read or offer this property. See examples, below.</p>
<p>The list or array will be converted to a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>
of the appropriate dtype, if necessary. Any <a class="reference external" href="https://numpy.org/doc/stable/reference/constants.html#numpy.nan" title="(in NumPy v1.26)"><code class="xref py py-data docutils literal notranslate"><span class="pre">numpy.nan</span></code></a> values
will converted to the appropriate missing value. The PLINK <a class="reference external" href="https://www.cog-genomics.org/plink2/formats#fam">.fam specification</a>
and <a class="reference external" href="https://www.cog-genomics.org/plink2/formats#bim">.bim specification</a>
lists the dtypes and missing values for each property.</p>
</div></blockquote>
</p></li>
<li><p><strong>count_A1</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – True (default) to count the number of A1 alleles (the PLINK standard).
False to count the number of A2 alleles.</p></li>
<li><p><strong>num_threads</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The number of threads with which to read data. Defaults to all available
processors.
Can also be set with these environment variables (listed in priority order):
‘PST_NUM_THREADS’, ‘NUM_THREADS’, ‘MKL_NUM_THREADS’.</p></li>
<li><p><strong>skip_format_check</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – False (default) to immediately check for expected starting bytes in
the .bed file. True to delay the check until (and if) data is read.</p></li>
<li><p><strong>fam_location</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em> or </em><em>URL</em><em>, </em><em>optional</em>) – Path to the file containing information about each individual (sample).
Defaults to replacing the .bed file’s suffix with .fam.</p></li>
<li><p><strong>bim_location</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><em>str URL</em><em>, </em><em>optional</em>) – Path to the file containing information about each SNP (variant).
Defaults to replacing the .bed file’s suffix with .bim.</p></li>
<li><p><strong>cloud_options</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a><em>, </em><em>optional</em>) – A dictionary of options for reading from cloud storage. The default is an empty.</p></li>
<li><p><strong>max_concurrent_requests</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The maximum number of concurrent requests to make to the cloud storage service.
Defaults to 10.</p></li>
<li><p><strong>max_chunk_bytes</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The maximum number of bytes to read in a single request to the cloud storage
service. Defaults to 8MB.</p></li>
<li><p><strong>filepath</strong> (<em>same as location</em>) – Deprecated. Use location instead.</p></li>
<li><p><strong>fam_filepath</strong> (<em>same as fam_location</em>) – Deprecated. Use fam_location instead.</p></li>
<li><p><strong>bim_filepath</strong> (<em>same as bim_location</em>) – Deprecated. Use bim_location instead.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>an open_bed object</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#bed_reader.open_bed" title="bed_reader.open_bed">open_bed</a></p>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>Open a local file and list individual (sample) <a class="reference internal" href="#bed_reader.open_bed.iid" title="bed_reader.open_bed.iid"><code class="xref py py-attr docutils literal notranslate"><span class="pre">iid</span></code></a> and SNP (variant) <a class="reference internal" href="#bed_reader.open_bed.sid" title="bed_reader.open_bed.sid"><code class="xref py py-attr docutils literal notranslate"><span class="pre">sid</span></code></a>. Then, <a class="reference internal" href="#bed_reader.open_bed.read" title="bed_reader.open_bed.read"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read()</span></code></a>
the whole file.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">iid</span><span class="p">)</span>
<span class="go">[&#39;iid1&#39; &#39;iid2&#39; &#39;iid3&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">sid</span><span class="p">)</span>
<span class="go">[&#39;sid1&#39; &#39;sid2&#39; &#39;sid3&#39; &#39;sid4&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="go">[[ 1.  0. nan  0.]</span>
<span class="go"> [ 2.  0. nan  2.]</span>
<span class="go"> [ 0.  1.  2.  0.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">bed</span>  <span class="c1"># optional: delete bed object</span>
</pre></div>
</div>
<p>Open a cloud file with a non-default timeout.
Then, read the data for one SNP (variant)
at index position 2.</p>
<p>See <a class="reference internal" href="cloud_urls.html"><span class="doc">Cloud URL Examples</span></a> for details on reading files from cloud storage.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="s2">&quot;https://raw.githubusercontent.com/fastlmm/bed-sample-files/main/small.bed&quot;</span><span class="p">,</span>
<span class="gp">... </span>              <span class="n">cloud_options</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;timeout&quot;</span><span class="p">:</span> <span class="s2">&quot;10s&quot;</span><span class="p">})</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="mi">2</span><span class="p">]))</span>
<span class="go">[[nan]</span>
<span class="go"> [nan]</span>
<span class="go"> [ 2.]]</span>
</pre></div>
</div>
<p>With the local file, replace <a class="reference internal" href="#bed_reader.open_bed.iid" title="bed_reader.open_bed.iid"><code class="xref py py-attr docutils literal notranslate"><span class="pre">iid</span></code></a>.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">properties</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;iid&quot;</span><span class="p">:[</span><span class="s2">&quot;sample1&quot;</span><span class="p">,</span><span class="s2">&quot;sample2&quot;</span><span class="p">,</span><span class="s2">&quot;sample3&quot;</span><span class="p">]})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">iid</span><span class="p">)</span> <span class="c1"># replaced</span>
<span class="go">[&#39;sample1&#39; &#39;sample2&#39; &#39;sample3&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">sid</span><span class="p">)</span> <span class="c1"># same as before</span>
<span class="go">[&#39;sid1&#39; &#39;sid2&#39; &#39;sid3&#39; &#39;sid4&#39;]</span>
</pre></div>
</div>
<p>Give the number of individuals (samples) and SNPs (variants) so that the .fam and
.bim files need never be opened.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">iid_count</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sid_count</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="go">[[ 1.  0. nan  0.]</span>
<span class="go"> [ 2.  0. nan  2.]</span>
<span class="go"> [ 0.  1.  2.  0.]]</span>
</pre></div>
</div>
<p>Mark some properties as “don’t read or offer”.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="n">properties</span><span class="o">=</span><span class="p">{</span>
<span class="gp">... </span>   <span class="s2">&quot;father&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;mother&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;sex&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;pheno&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
<span class="gp">... </span>   <span class="s2">&quot;allele_1&quot;</span> <span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;allele_2&quot;</span><span class="p">:</span><span class="kc">None</span> <span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">iid</span><span class="p">)</span>        <span class="c1"># read from file</span>
<span class="go">[&#39;iid1&#39; &#39;iid2&#39; &#39;iid3&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">allele_2</span><span class="p">)</span>   <span class="c1"># not read and not offered</span>
<span class="go">None</span>
</pre></div>
</div>
<p>See the <a class="reference internal" href="#bed_reader.open_bed.read" title="bed_reader.open_bed.read"><code class="xref py py-meth docutils literal notranslate"><span class="pre">read()</span></code></a> for details of reading batches via slicing and fancy indexing.</p>
<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.allele_1">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">allele_1</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.allele_1" title="Link to this definition"></a></dt>
<dd><p>First allele of each SNP (variant).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>If needed, will cause a one-time read of the .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">allele_1</span><span class="p">)</span>
<span class="go">[&#39;A&#39; &#39;T&#39; &#39;A&#39; &#39;T&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.allele_2">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">allele_2</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.allele_2" title="Link to this definition"></a></dt>
<dd><p>Second allele of each SNP (variant),</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>If needed, will cause a one-time read of the .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">allele_2</span><span class="p">)</span>
<span class="go">[&#39;A&#39; &#39;C&#39; &#39;C&#39; &#39;G&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.bp_position">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bp_position</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.bp_position" title="Link to this definition"></a></dt>
<dd><p>Base-pair position of each SNP (variant).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of int</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>0 represents a missing value.</p>
<p>If needed, will cause a one-time read of the .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">bp_position</span><span class="p">)</span>
<span class="go">[   1  100 1000 1004]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.chromosome">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.chromosome" title="Link to this definition"></a></dt>
<dd><p>Chromosome of each SNP (variant)</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>‘0’ represents a missing value.</p>
<p>If needed, will cause a one-time read of the .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
<span class="go">[&#39;1&#39; &#39;1&#39; &#39;5&#39; &#39;Y&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.cm_position">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">cm_position</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.cm_position" title="Link to this definition"></a></dt>
<dd><p>Centimorgan position of each SNP (variant).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of float</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>0.0 represents a missing value.</p>
<p>If needed, will cause a one-time read of the .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">cm_position</span><span class="p">)</span>
<span class="go">[ 100.4 2000.5 4000.7 7000.9]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.father">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">father</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.father" title="Link to this definition"></a></dt>
<dd><p>Father id of each individual (sample).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>‘0’ represents a missing value.</p>
<p>If needed, will cause a one-time read of the .fam file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">father</span><span class="p">)</span>
<span class="go">[&#39;iid23&#39; &#39;iid23&#39; &#39;iid22&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.fid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">fid</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.fid" title="Link to this definition"></a></dt>
<dd><p>Family id of each individual (sample).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>‘0’ represents a missing value.</p>
<p>If needed, will cause a one-time read of the .fam file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">fid</span><span class="p">)</span>
<span class="go">[&#39;fid1&#39; &#39;fid1&#39; &#39;fid2&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.iid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">iid</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.iid" title="Link to this definition"></a></dt>
<dd><p>Individual id of each individual (sample).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>If needed, will cause a one-time read of the .fam file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">iid</span><span class="p">)</span>
<span class="go">[&#39;iid1&#39; &#39;iid2&#39; &#39;iid3&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.iid_count">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">iid_count</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.iid_count" title="Link to this definition"></a></dt>
<dd><p>Number of individuals (samples).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>number of individuals</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)">int</a></p>
</dd>
</dl>
<p>If needed, will cause a fast line-count of the .fam file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">iid_count</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.major">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">major</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></em><a class="headerlink" href="#bed_reader.open_bed.major" title="Link to this definition"></a></dt>
<dd><p>Major mode of a local .bed file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>‘SNP’ or ‘individual’</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)">str</a></p>
</dd>
</dl>
<p>Almost all PLINK 1.9 .bed files are ‘SNP’ major. This makes
reading the data by SNP(s) fast.</p>
<section id="errors">
<h4>Errors<a class="headerlink" href="#errors" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>ValueError</dt><dd><p>If the file is a cloud file.</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">major</span><span class="p">)</span>
<span class="go">SNP</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.mother">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mother</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.mother" title="Link to this definition"></a></dt>
<dd><p>Mother id of each individual (sample).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>‘0’ represents a missing value.</p>
<p>If needed, will cause a one-time read of the .fam file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">mother</span><span class="p">)</span>
<span class="go">[&#39;iid34&#39; &#39;iid34&#39; &#39;iid33&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.pheno">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pheno</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.pheno" title="Link to this definition"></a></dt>
<dd><p>A phenotype for each individual (sample)
(seldom used).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>‘0’ may represent a missing value.</p>
<p>If needed, will cause a one-time read of the .fam file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">pheno</span><span class="p">)</span>
<span class="go">[&#39;red&#39; &#39;red&#39; &#39;blue&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.properties">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">properties</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">array</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#bed_reader.open_bed.properties" title="Link to this definition"></a></dt>
<dd><p>All the properties returned as a dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>all the properties</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)">dict</a></p>
</dd>
</dl>
<p>The keys of the dictionary are the names of the properties, namely:</p>
<blockquote>
<div><p>“fid” (family id), “iid” (individual or sample id), “father” (father id),
“mother” (mother id), “sex”, “pheno” (phenotype), “chromosome”, “sid”
(SNP or variant id), “cm_position” (centimorgan position), “bp_position”
(base-pair position), “allele_1”, “allele_2”.</p>
</div></blockquote>
<p>The values are <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></a>.</p>
<p>If needed, will cause a one-time read of the .fam and .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">properties</span><span class="p">))</span> <span class="c1">#length of dict</span>
<span class="go">12</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bed_reader.open_bed.property_item">
<span class="sig-name descname"><span class="pre">property_item</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></span></span><a class="reference internal" href="_modules/bed_reader/_open_bed.html#open_bed.property_item"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.open_bed.property_item" title="Link to this definition"></a></dt>
<dd><p>Retrieve one property by name.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a property value</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>The name is one of these:</p>
<blockquote>
<div><p>“fid” (family id), “iid” (individual or sample id), “father” (father id),
“mother” (mother id), “sex”, “pheno” (phenotype), “chromosome”, “sid”
(SNP or variant id), “cm_position” (centimorgan position), “bp_position”
(base-pair position), “allele_1”, “allele_2”.</p>
</div></blockquote>
<p>If needed, will cause a one-time read of the .fam or .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">property_item</span><span class="p">(</span><span class="s1">&#39;chromosome&#39;</span><span class="p">))</span>
<span class="go">[&#39;1&#39; &#39;1&#39; &#39;5&#39; &#39;Y&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bed_reader.open_bed.read">
<span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#type" title="(in Python v3.12)"><span class="pre">type</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'float32'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'F'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_python_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_concurrent_requests</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_chunk_bytes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></span></span><a class="reference internal" href="_modules/bed_reader/_open_bed.html#open_bed.read"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.open_bed.read" title="Link to this definition"></a></dt>
<dd><p>Read genotype information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> – <p>An optional expression specifying the individuals (samples) and SNPs
(variants) to read. (See examples, below).
Defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>, meaning read all.</p>
<p>(If index is a tuple, the first component indexes the individuals and the
second indexes
the SNPs. If it is not a tuple and not None, it indexes SNPs.)</p>
</p></li>
<li><p><strong>dtype</strong> (<em>{'float32'</em><em> (</em><em>default</em><em>)</em><em>, </em><em>'float64'</em><em>, </em><em>'int8'}</em><em>, </em><em>optional</em>) – The desired data-type for the returned array.</p></li>
<li><p><strong>order</strong> (<em>{'F'</em><em>,</em><em>'C'}</em><em>, </em><em>optional</em>) – The desired memory layout for the returned array.
Defaults to <code class="docutils literal notranslate"><span class="pre">F</span></code> (Fortran order, which is SNP-major).</p></li>
<li><p><strong>force_python_only</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – If False (default), uses the faster Rust code; otherwise it uses the slower
pure Python code.</p></li>
<li><p><strong>num_threads</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The number of threads with which to read data. Defaults to all available
processors.
Can also be set with <a class="reference internal" href="#bed_reader.open_bed" title="bed_reader.open_bed"><code class="xref py py-class docutils literal notranslate"><span class="pre">open_bed</span></code></a> or these
environment variables (listed in priority order):
‘PST_NUM_THREADS’, ‘NUM_THREADS’, ‘MKL_NUM_THREADS’.</p></li>
<li><p><strong>max_concurrent_requests</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The maximum number of concurrent requests to make to the cloud storage
service. Defaults to 10.</p></li>
<li><p><strong>max_chunk_bytes</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The maximum number of bytes to read in a single request to the cloud
storage service. Defaults to 8MB.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>2-D array containing values of 0, 1, 2, or missing</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>Rows represent individuals (samples). Columns represent SNPs (variants).</p>
<p>For <code class="docutils literal notranslate"><span class="pre">dtype</span></code> ‘float32’ and ‘float64’, NaN indicates missing values.
For ‘int8’, -127 indicates missing values.</p>
<p class="rubric">Examples</p>
<p>To read all data in a .bed file, set <code class="docutils literal notranslate"><span class="pre">index</span></code> to <code class="docutils literal notranslate"><span class="pre">None</span></code>. This is the default.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">())</span>
<span class="go">[[ 1.  0. nan  0.]</span>
<span class="go"> [ 2.  0. nan  2.]</span>
<span class="go"> [ 0.  1.  2.  0.]]</span>
</pre></div>
</div>
<p>To read selected individuals (samples) and/or SNPs (variants), set each part of
a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.s_.html#numpy.s_" title="(in NumPy v1.26)"><code class="xref py py-data docutils literal notranslate"><span class="pre">numpy.s_</span></code></a> to an <cite>int</cite>, a list of <cite>int</cite>, a slice expression, or
a list of <cite>bool</cite>.
Negative integers count from the end of the list.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="mi">2</span><span class="p">]))</span>  <span class="c1"># read the SNPs indexed by 2.</span>
<span class="go">[[nan]</span>
<span class="go"> [nan]</span>
<span class="go"> [ 2.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,[</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">0</span><span class="p">]]))</span>  <span class="c1"># read the SNPs indexed by 2, 3, and 0</span>
<span class="go">[[nan  0.  1.]</span>
<span class="go"> [nan  2.  2.]</span>
<span class="go"> [ 2.  0.  0.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># read SNPs from 1 (inclusive) to 4 (exclusive)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="mi">1</span><span class="p">:</span><span class="mi">4</span><span class="p">]))</span>
<span class="go">[[ 0. nan  0.]</span>
<span class="go"> [ 0. nan  2.]</span>
<span class="go"> [ 1.  2.  0.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span><span class="p">))</span> <span class="c1"># print unique chrom values</span>
<span class="go">[&#39;1&#39; &#39;5&#39; &#39;Y&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span><span class="o">==</span><span class="s1">&#39;5&#39;</span><span class="p">]))</span> <span class="c1"># read all SNPs in chrom 5</span>
<span class="go">[[nan]</span>
<span class="go"> [nan]</span>
<span class="go"> [ 2.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[</span><span class="mi">0</span><span class="p">,:]))</span> <span class="c1"># Read 1st individual (across all SNPs)</span>
<span class="go">[[ 1.  0. nan  0.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[::</span><span class="mi">2</span><span class="p">,:]))</span> <span class="c1"># Read every 2nd individual</span>
<span class="go">[[ 1.  0. nan  0.]</span>
<span class="go"> [ 0.  1.  2.  0.]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#read last and 2nd-to-last individuals and the last SNPs</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">2</span><span class="p">],</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>
<span class="go">[[0.]</span>
<span class="go"> [2.]]</span>
</pre></div>
</div>
<p>You can give a dtype for the output.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;int8&#39;</span><span class="p">))</span>
<span class="go">[[   1    0 -127    0]</span>
<span class="go"> [   2    0 -127    2]</span>
<span class="go"> [   0    1    2    0]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">del</span> <span class="n">bed</span>  <span class="c1"># optional: delete bed object</span>
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bed_reader.open_bed.read_sparse">
<span class="sig-name descname"><span class="pre">read_sparse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#type" title="(in Python v3.12)"><span class="pre">type</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'float32'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'csc'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_concurrent_requests</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_chunk_bytes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix" title="(in SciPy v1.13.0)"><span class="pre">csc_matrix</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix" title="(in SciPy v1.13.0)"><span class="pre">csr_matrix</span></a></span></span><a class="reference internal" href="_modules/bed_reader/_open_bed.html#open_bed.read_sparse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.open_bed.read_sparse" title="Link to this definition"></a></dt>
<dd><p>Read genotype information into a <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/sparse.html#module-scipy.sparse" title="(in SciPy v1.13.0)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scipy.sparse</span></code></a> matrix. Sparse matrices
may be useful when the data is mostly zeros.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This method requires <a class="reference external" href="https://docs.scipy.org/doc/scipy/index.html#module-scipy" title="(in SciPy v1.13.0)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">scipy</span></code></a>. Install <cite>scipy</cite> with:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>--upgrade<span class="w"> </span>bed-reader<span class="o">[</span>sparse<span class="o">]</span>
</pre></div>
</div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>index</strong> – <p>An optional expression specifying the individuals (samples) and SNPs
(variants) to read. (See examples, below).
Defaults to <code class="docutils literal notranslate"><span class="pre">None</span></code>, meaning read all.</p>
<p>(If index is a tuple, the first component indexes the individuals and the
second indexes
the SNPs. If it is not a tuple and not None, it indexes SNPs.)</p>
</p></li>
<li><p><strong>dtype</strong> (<em>{'float32'</em><em> (</em><em>default</em><em>)</em><em>, </em><em>'float64'</em><em>, </em><em>'int8'}</em><em>, </em><em>optional</em>) – The desired data-type for the returned array.</p></li>
<li><p><strong>batch_size</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Number of dense columns or rows to read at a time, internally.
Defaults to round(sqrt(total-number-of-columns-or-rows-to-read)).</p></li>
<li><p><strong>format</strong> (<em>{'csc'</em><em>,</em><em>'csr'}</em><em>, </em><em>optional</em>) – The desired format of the sparse matrix.
Defaults to <code class="docutils literal notranslate"><span class="pre">csc</span></code> (Compressed Sparse Column, which is SNP-major).</p></li>
<li><p><strong>num_threads</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The number of threads with which to read data. Defaults to all available
processors.
Can also be set with <a class="reference internal" href="#bed_reader.open_bed" title="bed_reader.open_bed"><code class="xref py py-class docutils literal notranslate"><span class="pre">open_bed</span></code></a> or these
environment variables (listed in priority order):
‘PST_NUM_THREADS’, ‘NUM_THREADS’, ‘MKL_NUM_THREADS’.</p></li>
<li><p><strong>max_concurrent_requests</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The maximum number of concurrent requests to make to the cloud storage
service. Defaults to 10.</p></li>
<li><p><strong>max_chunk_bytes</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The maximum number of bytes to read in a single request to the cloud
storage service. Defaults to 8MB.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>a <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix" title="(in SciPy v1.13.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">scipy.sparse.csc_matrix</span></code></a> (default) or <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html#scipy.sparse.csr_matrix" title="(in SciPy v1.13.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">scipy.sparse.csr_matrix</span></code></a></p></li>
<li><p><em>Rows represent individuals (samples). Columns represent SNPs (variants).</em></p></li>
<li><p>For <code class="docutils literal notranslate"><span class="pre">dtype</span></code> ‘float32’ and ‘float64’, NaN indicates missing values.</p></li>
<li><p><em>For ‘int8’, -127 indicates missing values.</em></p></li>
</ul>
</p>
</dd>
</dl>
<p>The memory used by the final sparse matrix is approximately:</p>
<blockquote>
<div><p># of non-zero values * (4 bytes + 1 byte (for int8))</p>
</div></blockquote>
<p>For example, consider reading 1000 individuals (samples) x 50,000 SNPs (variants)
into csc format where the data is 97% sparse.
The memory used will be about 7.5 MB (1000 x 50,000 x 3% x 5 bytes).
This is 15% of the 50 MB needed by a dense matrix.</p>
<p>Internally, the function reads the data via small dense matrices.
For this example, by default, the function will read 1000 individuals x 224 SNPs
(because 224 * 224 is about 50,000).
The memory used by the small dense matrix is 1000 x 244 x 1 byte (for int8) = 0.224 MB.</p>
<p>You can set <cite>batch_size</cite>. Larger values will be faster.
Smaller values will use less memory.</p>
<p>For this example, we might want to set the <cite>batch_size</cite> to 5000. Then,
the memory used by the small dense matrix
would be 1000 x 5000 x 1 byte (for int8) = 5 MB,
similar to the 7.5 MB needed for the final sparse matrix.</p>
<p class="rubric">Examples</p>
<p>Read all data in a .bed file into a <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_matrix.html#scipy.sparse.csc_matrix" title="(in SciPy v1.13.0)"><code class="xref py py-class docutils literal notranslate"><span class="pre">scipy.sparse.csc_matrix</span></code></a>.
The file has 10 individuals (samples) by 20 SNPs (variants).
All but eight values are 0.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># pip install bed-reader[samples,sparse]  # if needed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;sparse.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">val_sparse</span> <span class="o">=</span> <span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">val_sparse</span><span class="p">)</span> 
<span class="go">(10, 20)</span>
<span class="go">    (8, 4)  1</span>
<span class="go">    (8, 5)  2</span>
<span class="go">    (0, 8)  2</span>
<span class="go">    (4, 9)  1</span>
<span class="go">    (7, 9)  1</span>
<span class="go">    (5, 11) 1</span>
<span class="go">    (2, 12) 1</span>
<span class="go">    (3, 12) 1</span>
</pre></div>
</div>
<p>To read selected individuals (samples) and/or SNPs (variants), set each part of
a <a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.s_.html#numpy.s_" title="(in NumPy v1.26)"><code class="xref py py-data docutils literal notranslate"><span class="pre">numpy.s_</span></code></a> to an <cite>int</cite>, a list of <cite>int</cite>, a slice expression, or
a list of <cite>bool</cite>.
Negative integers count from the end of the list.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bed</span> <span class="o">=</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="mi">5</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">))</span>  <span class="c1"># read the SNPs indexed by 5. </span>
<span class="go">(8, 0)    2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># read the SNPs indexed by 5, 4, and 0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,[</span><span class="mi">5</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">0</span><span class="p">]],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">))</span> 
<span class="go">(8, 0)      2</span>
<span class="go">(8, 1)      1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># read SNPs from 1 (inclusive) to 11 (exclusive)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="mi">1</span><span class="p">:</span><span class="mi">11</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">))</span> 
<span class="go">(8, 3)      1</span>
<span class="go">(8, 4)      2</span>
<span class="go">(0, 7)      2</span>
<span class="go">(4, 8)      1</span>
<span class="go">(7, 8)      1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span><span class="p">))</span> <span class="c1"># print unique chrom values</span>
<span class="go">[&#39;1&#39; &#39;5&#39; &#39;Y&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># read all SNPs in chrom 5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[:,</span><span class="n">bed</span><span class="o">.</span><span class="n">chromosome</span><span class="o">==</span><span class="s1">&#39;5&#39;</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">))</span> 
<span class="go">(8, 0)      1</span>
<span class="go">(8, 1)      2</span>
<span class="go">(0, 4)      2</span>
<span class="go">(4, 5)      1</span>
<span class="go">(7, 5)      1</span>
<span class="go">(5, 7)      1</span>
<span class="go">(2, 8)      1</span>
<span class="go">(3, 8)      1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Read 1st individual (across all SNPs)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[</span><span class="mi">0</span><span class="p">,:],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">))</span> 
<span class="go">(0, 8)      2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[::</span><span class="mi">2</span><span class="p">,:],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">))</span> <span class="c1"># Read every 2nd individual </span>
<span class="go">(4, 4)    1</span>
<span class="go">(4, 5)    2</span>
<span class="go">(0, 8)    2</span>
<span class="go">(2, 9)    1</span>
<span class="go">(1, 12)   1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># read last and 2nd-to-last individuals and the 15th-from-the-last SNP</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">read_sparse</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">s_</span><span class="p">[[</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">2</span><span class="p">],</span><span class="o">-</span><span class="mi">15</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="s2">&quot;int8&quot;</span><span class="p">))</span> 
<span class="go">(1, 0)      2</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.sex">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sex</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.sex" title="Link to this definition"></a></dt>
<dd><p>Sex of each individual (sample).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of 0, 1, or 2</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>0 is unknown, 1 is male, 2 is female</p>
<p>If needed, will cause a one-time read of the .fam file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">sex</span><span class="p">)</span>
<span class="go">[1 2 0]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.shape">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">shape</span></span><a class="headerlink" href="#bed_reader.open_bed.shape" title="Link to this definition"></a></dt>
<dd><p>Number of individuals (samples) and SNPs (variants).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>number of individuals, number of SNPs</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)">int</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)">int</a>)</p>
</dd>
</dl>
<p>If needed, will cause a fast line-count of the .fam and .bim files.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">(3, 4)</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.sid">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sid</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.sid" title="Link to this definition"></a></dt>
<dd><p>SNP id of each SNP (variant).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>array of str</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)">numpy.ndarray</a></p>
</dd>
</dl>
<p>If needed, will cause a one-time read of the .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">sid</span><span class="p">)</span>
<span class="go">[&#39;sid1&#39; &#39;sid2&#39; &#39;sid3&#39; &#39;sid4&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="bed_reader.open_bed.sid_count">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sid_count</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></em><a class="headerlink" href="#bed_reader.open_bed.sid_count" title="Link to this definition"></a></dt>
<dd><p>Number of SNPs (variants).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>number of SNPs</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)">int</a></p>
</dd>
</dl>
<p>If needed, will cause a fast line-count of the .bim file.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span><span class="p">,</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">file_name</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed</span><span class="o">.</span><span class="n">sid_count</span><span class="p">)</span>
<span class="go">4</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

</section>
<section id="to-bed">
<h3>to_bed<a class="headerlink" href="#to-bed" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="bed_reader.to_bed">
<span class="sig-prename descclassname"><span class="pre">bed_reader.</span></span><span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">val</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.26)"><span class="pre">ndarray</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">properties</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_A1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fam_filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bim_filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">major</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'SNP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_python_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bed_reader/_to_bed.html#to_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.to_bed" title="Link to this definition"></a></dt>
<dd><p>Write values to a file in PLINK .bed format.</p>
<p>If your data is too large to fit in memory, use <a class="reference internal" href="#bed_reader.create_bed" title="bed_reader.create_bed"><code class="xref py py-class docutils literal notranslate"><span class="pre">create_bed</span></code></a> instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filepath</strong> – .bed file to write to.</p></li>
<li><p><strong>val</strong> (<em>array-like:</em>) – A two-dimension array (or array-like object) of values. The values should
be (or be convertible to) all floats or all integers.
The values should be 0, 1, 2, or missing.
If floats, missing is <code class="docutils literal notranslate"><span class="pre">np.nan</span></code>. If integers, missing is -127.</p></li>
<li><p><strong>properties</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a><em>, </em><em>optional</em>) – <p>A dictionary of property names and values to write to the .fam and .bim files.
Any properties not mentioned will be filled in with default values.</p>
<p>The possible property names are:</p>
<blockquote>
<div><blockquote>
<div><p>”fid” (family id), “iid” (individual or sample id), “father” (father id),
“mother” (mother id), “sex”, “pheno” (phenotype), “chromosome”, “sid”
(SNP or variant id), “cm_position” (centimorgan position), “bp_position”
(base-pair position), “allele_1”, “allele_2”.</p>
</div></blockquote>
<p>The values are lists or arrays. See example, below.</p>
</div></blockquote>
</p></li>
<li><p><strong>count_A1</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – True (default) to count the number of A1 alleles (the PLINK standard).
False to count the number of A2 alleles.</p></li>
<li><p><strong>fam_filepath</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – Path to the file containing information about each individual (sample).
Defaults to replacing the .bed file’s suffix with .fam.</p></li>
<li><p><strong>bim_filepath</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – Path to the file containing information about each SNP (variant).
Defaults to replacing the .bed file’s suffix with .bim.</p></li>
<li><p><strong>major</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – Use “SNP” (default) to write the file is usual SNP-major mode. This makes reading
the data SNP-by-SNP faster. Use “individual” to write the file in the uncommon
individual-major mode.</p></li>
<li><p><strong>force_python_only</strong> – If False (default), uses the faster Rust helper functions; otherwise it uses the slower
pure Python code.</p></li>
<li><p><strong>num_threads</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – The number of threads with which to write data.
Defaults to all available processors.
Can also be set with these environment variables (listed in priority order):
‘PST_NUM_THREADS’, ‘NUM_THREADS’, ‘MKL_NUM_THREADS’.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<p>In this example, all properties are given.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">to_bed</span><span class="p">,</span> <span class="n">tmp_path</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_file</span> <span class="o">=</span> <span class="n">tmp_path</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;small.bed&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">val</span> <span class="o">=</span> <span class="p">[[</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],</span>
<span class="gp">... </span>       <span class="p">[</span><span class="mf">2.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">],</span>
<span class="gp">... </span>       <span class="p">[</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>   <span class="s2">&quot;fid&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;fid1&quot;</span><span class="p">,</span> <span class="s2">&quot;fid1&quot;</span><span class="p">,</span> <span class="s2">&quot;fid2&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;iid&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;iid1&quot;</span><span class="p">,</span> <span class="s2">&quot;iid2&quot;</span><span class="p">,</span> <span class="s2">&quot;iid3&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;father&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;iid23&quot;</span><span class="p">,</span> <span class="s2">&quot;iid23&quot;</span><span class="p">,</span> <span class="s2">&quot;iid22&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;mother&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;iid34&quot;</span><span class="p">,</span> <span class="s2">&quot;iid34&quot;</span><span class="p">,</span> <span class="s2">&quot;iid33&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;pheno&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;Y&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;sid&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sid1&quot;</span><span class="p">,</span> <span class="s2">&quot;sid2&quot;</span><span class="p">,</span> <span class="s2">&quot;sid3&quot;</span><span class="p">,</span> <span class="s2">&quot;sid4&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;cm_position&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">100.4</span><span class="p">,</span> <span class="mf">2000.5</span><span class="p">,</span> <span class="mf">4000.7</span><span class="p">,</span> <span class="mf">7000.9</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;bp_position&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">1004</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;allele_1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;allele_2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">],</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">to_bed</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">val</span><span class="p">,</span> <span class="n">properties</span><span class="o">=</span><span class="n">properties</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, no properties are given, so default values are assigned.
If we then read the new file and list the chromosome property,
it is an array of ‘0’s, the default chromosome value.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">output_file2</span> <span class="o">=</span> <span class="n">tmp_path</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;small2.bed&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">val</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">127</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">127</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">to_bed</span><span class="p">(</span><span class="n">output_file2</span><span class="p">,</span> <span class="n">val</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">output_file2</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed2</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed2</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
<span class="go">[&#39;0&#39; &#39;0&#39; &#39;0&#39; &#39;0&#39;]</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="create-bed">
<h3>create_bed<a class="headerlink" href="#create-bed" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="bed_reader.create_bed">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">bed_reader.</span></span><span class="sig-name descname"><span class="pre">create_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">iid_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">sid_count</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">properties</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.List" title="(in Python v3.12)"><span class="pre">List</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">count_A1</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fam_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bim_location</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">major</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'SNP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_python_only</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_threads</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bed_reader/_to_bed.html#create_bed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.create_bed" title="Link to this definition"></a></dt>
<dd><p>Open a file to write values into PLINK .bed format.</p>
<p>Values may be given in a SNP-by-SNP (or individual-by-individual) manner.
For large datasets, <cite>create_bed</cite> requires less memory than <a class="reference internal" href="#bed_reader.to_bed" title="bed_reader.to_bed"><code class="xref py py-func docutils literal notranslate"><span class="pre">to_bed()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>location</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – local .bed file to create.</p></li>
<li><p><strong>iid_count</strong> (<em>int:</em>) – The number of individuals (samples).</p></li>
<li><p><strong>sid_count</strong> (<em>int:</em>) – The number of SNPs (variants).</p></li>
<li><p><strong>properties</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.12)"><em>dict</em></a><em>, </em><em>optional</em>) – <p>A dictionary of property names and values to write to the .fam and .bim files.
Any properties not mentioned will be filled in with default values.</p>
<p>The possible property names are:</p>
<blockquote>
<div><blockquote>
<div><p>”fid” (family id), “iid” (individual or sample id), “father” (father id),
“mother” (mother id), “sex”, “pheno” (phenotype), “chromosome”, “sid”
(SNP or variant id), “cm_position” (centimorgan position), “bp_position”
(base-pair position), “allele_1”, “allele_2”.</p>
</div></blockquote>
<p>The values are lists or arrays. See example, below.</p>
</div></blockquote>
</p></li>
<li><p><strong>count_A1</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>, </em><em>optional</em>) – True (default) to count the number of A1 alleles (the PLINK standard).
False to count the number of A2 alleles.</p></li>
<li><p><strong>fam_location</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – Path to the file containing information about each individual (sample).
Defaults to replacing the .bed file’s suffix with .fam.</p></li>
<li><p><strong>bim_location</strong> (<a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><em>pathlib.Path</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – Path to the file containing information about each SNP (variant).
Defaults to replacing the .bed file’s suffix with .bim.</p></li>
<li><p><strong>major</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><em>optional</em>) – Use “SNP” (default) to write the file is usual SNP-major mode. This makes reading
the data SNP-by-SNP faster. Use “individual” to write the file in the uncommon
individual-major mode.</p></li>
<li><p><strong>force_python_only</strong> – If False (default), uses the faster Rust helper functions; otherwise it uses the slower
pure Python code.</p></li>
<li><p><strong>num_threads</strong> (<em>None</em><em> or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a><em>, </em><em>optional</em>) – Not currently used.</p></li>
</ul>
</dd>
</dl>
<section id="create-bed-errors">
<h4>create_bed Errors<a class="headerlink" href="#create-bed-errors" title="Link to this heading"></a></h4>
<p>Raises an error if you write the wrong number of vectors or if any vector has the wrong length.</p>
<p>Also, all vector values must be 0, 1, 2, or missing. If floats, missing is <code class="docutils literal notranslate"><span class="pre">np.nan</span></code>. If integers, missing is -127.</p>
<p>Behind the scenes, <cite>create_bed</cite> first creates a temporary bed file. At the end, if there are no errors,
it renames the temporary file to the final file name. This helps prevent creation
of corrupted files.</p>
<p class="rubric">Examples</p>
<p>In this example, all properties are given and we write
the data out SNP-by-SNP. The data is floats.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">create_bed</span><span class="p">,</span> <span class="n">tmp_path</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_file</span> <span class="o">=</span> <span class="n">tmp_path</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;small.bed&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">properties</span> <span class="o">=</span> <span class="p">{</span>
<span class="gp">... </span>   <span class="s2">&quot;fid&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;fid1&quot;</span><span class="p">,</span> <span class="s2">&quot;fid1&quot;</span><span class="p">,</span> <span class="s2">&quot;fid2&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;iid&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;iid1&quot;</span><span class="p">,</span> <span class="s2">&quot;iid2&quot;</span><span class="p">,</span> <span class="s2">&quot;iid3&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;father&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;iid23&quot;</span><span class="p">,</span> <span class="s2">&quot;iid23&quot;</span><span class="p">,</span> <span class="s2">&quot;iid22&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;mother&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;iid34&quot;</span><span class="p">,</span> <span class="s2">&quot;iid34&quot;</span><span class="p">,</span> <span class="s2">&quot;iid33&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;sex&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;pheno&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="s2">&quot;blue&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;chromosome&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;5&quot;</span><span class="p">,</span> <span class="s2">&quot;Y&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;sid&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sid1&quot;</span><span class="p">,</span> <span class="s2">&quot;sid2&quot;</span><span class="p">,</span> <span class="s2">&quot;sid3&quot;</span><span class="p">,</span> <span class="s2">&quot;sid4&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;cm_position&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">100.4</span><span class="p">,</span> <span class="mf">2000.5</span><span class="p">,</span> <span class="mf">4000.7</span><span class="p">,</span> <span class="mf">7000.9</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;bp_position&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1000</span><span class="p">,</span> <span class="mi">1004</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;allele_1&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">],</span>
<span class="gp">... </span>   <span class="s2">&quot;allele_2&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">],</span>
<span class="gp">... </span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">create_bed</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">iid_count</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sid_count</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">properties</span><span class="o">=</span><span class="n">properties</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed_writer</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">bed_writer</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">])</span>
<span class="gp">... </span>    <span class="n">bed_writer</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">])</span>
<span class="gp">... </span>    <span class="n">bed_writer</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">])</span>
<span class="gp">... </span>    <span class="n">bed_writer</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="mf">0.0</span><span class="p">,</span> <span class="mf">2.0</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">])</span>
</pre></div>
</div>
<p>In this next example, no properties are given, so default values are assigned.
Also, we write out ints. Finally, we write the same data out as above,
but individual-by-individual.
If we then read the new file and list the chromosome property,
it is an array of ‘0’s, the default chromosome value.</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">output_file2</span> <span class="o">=</span> <span class="n">tmp_path</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;small2.bed&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">create_bed</span><span class="p">(</span><span class="n">output_file2</span><span class="p">,</span> <span class="n">iid_count</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">sid_count</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">major</span><span class="o">=</span><span class="s2">&quot;individual&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed_writer</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">bed_writer</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">127</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="gp">... </span>    <span class="n">bed_writer</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">127</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="gp">... </span>    <span class="n">bed_writer</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">open_bed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">open_bed</span><span class="p">(</span><span class="n">output_file2</span><span class="p">)</span> <span class="k">as</span> <span class="n">bed2</span><span class="p">:</span>
<span class="gp">... </span>    <span class="nb">print</span><span class="p">(</span><span class="n">bed2</span><span class="o">.</span><span class="n">chromosome</span><span class="p">)</span>
<span class="go">[&#39;0&#39; &#39;0&#39; &#39;0&#39; &#39;0&#39;]</span>
</pre></div>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="bed_reader.create_bed.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/bed_reader/_to_bed.html#create_bed.close"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.create_bed.close" title="Link to this definition"></a></dt>
<dd><p>Close the bed_writer, writing the file to disk. If you use
<a class="reference internal" href="#bed_reader.create_bed" title="bed_reader.create_bed"><code class="xref py py-class docutils literal notranslate"><span class="pre">create_bed</span></code></a> with the <cite>with</cite> statement,
you don’t need to use this.</p>
<p>See <a class="reference internal" href="#bed_reader.create_bed" title="bed_reader.create_bed"><code class="xref py py-class docutils literal notranslate"><span class="pre">create_bed</span></code></a> for more information.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="bed_reader.create_bed.write">
<span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vector</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/bed_reader/_to_bed.html#create_bed.write"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.create_bed.write" title="Link to this definition"></a></dt>
<dd><p>Write a vector of values to the bed_writer.</p>
<p>See <a class="reference internal" href="#bed_reader.create_bed" title="bed_reader.create_bed"><code class="xref py py-class docutils literal notranslate"><span class="pre">create_bed</span></code></a> for more information.</p>
</dd></dl>

</section>
</dd></dl>

</section>
<section id="sample-file">
<h3>sample_file<a class="headerlink" href="#sample-file" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="bed_reader.sample_file">
<span class="sig-prename descclassname"><span class="pre">bed_reader.</span></span><span class="sig-name descname"><span class="pre">sample_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filepath</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></span><a class="reference internal" href="_modules/bed_reader/_sample_data.html#sample_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.sample_file" title="Link to this definition"></a></dt>
<dd><p>Retrieve a sample .bed file. (Also retrieves associated .fam and .bim files).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filepath</strong> – Name of the sample .bed file.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Local name of sample .bed file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)">str</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function requires the <code class="xref py py-mod docutils literal notranslate"><span class="pre">pooch</span></code> package. Install <cite>pooch</cite> with:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>--upgrade<span class="w"> </span>bed-reader<span class="o">[</span>samples<span class="o">]</span>
</pre></div>
</div>
</div>
<p>By default this function puts files under the user’s cache directory.
Override this by setting
the <cite>BED_READER_DATA_DIR</cite> environment variable.</p>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># pip install bed-reader[samples]  # if needed</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">sample_file</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">file_name</span> <span class="o">=</span> <span class="n">sample_file</span><span class="p">(</span><span class="s2">&quot;small.bed&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;The local file name is &#39;</span><span class="si">{</span><span class="n">file_name</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>
<span class="go">The local file name is &#39;...small.bed&#39;</span>
</pre></div>
</div>
</dd></dl>

</section>
<section id="tmp-path">
<h3>tmp_path<a class="headerlink" href="#tmp-path" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="bed_reader.tmp_path">
<span class="sig-prename descclassname"><span class="pre">bed_reader.</span></span><span class="sig-name descname"><span class="pre">tmp_path</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><span class="pre">Path</span></a></span></span><a class="reference internal" href="_modules/bed_reader/_sample_data.html#tmp_path"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#bed_reader.tmp_path" title="Link to this definition"></a></dt>
<dd><p>Return a <a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">pathlib.Path</span></code></a> to a temporary directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a temporary directory</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/pathlib.html#pathlib.Path" title="(in Python v3.12)">pathlib.Path</a></p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-pycon notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">bed_reader</span> <span class="kn">import</span> <span class="n">to_bed</span><span class="p">,</span> <span class="n">tmp_path</span>
<span class="gp">&gt;&gt;&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_file</span> <span class="o">=</span> <span class="n">tmp_path</span><span class="p">()</span> <span class="o">/</span> <span class="s2">&quot;small3.bed&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">val</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">127</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">127</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">to_bed</span><span class="p">(</span><span class="n">output_file</span><span class="p">,</span> <span class="n">val</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

</section>
</section>
<section id="environment-variables">
<h2>Environment Variables<a class="headerlink" href="#environment-variables" title="Link to this heading"></a></h2>
<p>By default <a class="reference internal" href="#bed_reader.sample_file" title="bed_reader.sample_file"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sample_file()</span></code></a> puts files under the user’s cache directory. Override this by setting
the <code class="docutils literal notranslate"><span class="pre">BED_READER_DATA_DIR</span></code> environment variable.</p>
<p>By default, <a class="reference internal" href="#bed_reader.open_bed" title="bed_reader.open_bed"><code class="xref py py-class docutils literal notranslate"><span class="pre">open_bed</span></code></a> uses all available processors. Override this with the <code class="docutils literal notranslate"><span class="pre">num_threads</span></code>
parameter or by setting environment variable (listed in priority order):
‘PST_NUM_THREAD</p>
</section>
<section id="cloud-url-examples">
<h2>Cloud URL Examples<a class="headerlink" href="#cloud-url-examples" title="Link to this heading"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="cloud_urls.html">Cloud URL Examples</a></li>
</ul>
</div>
</section>
<section id="indices-and-tables">
<h2>Indices and Tables<a class="headerlink" href="#indices-and-tables" title="Link to this heading"></a></h2>
<blockquote>
<div><ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div></blockquote>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="cloud_urls.html" class="btn btn-neutral float-right" title="Cloud URL Examples" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Carl Kadie.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>